api/host_test.go

```diff
@@ -1,5 +1,485 @@
+	}
+	defer st.server.Close()
+
+	// Announce the host and start accepting contracts.
+	if err := st.announceHost(); err != nil {
+		t.Fatal(err)
+	}
+	if err := st.acceptContracts(); err != nil {
+		t.Fatal(err)
+	}
+	if err := st.setHostStorage(); err != nil {
+		t.Fatal(err)
+	}
+
+	// Find out how large the host's initial storage folder is.
+	var sg StorageGET
+	if err := st.getAPI("/host/storage", &sg); err != nil {
+		t.Fatal(err)
+	}
+	defaultSize := sg.Folders[0].Capacity
+	// Convert defaultSize (uint64) to a string for the API call.
+	defaultSizeString := strconv.FormatUint(defaultSize, 10)
+
+	resizeValues := url.Values{}
+	resizeValues.Set("path", st.dir)
+	resizeValues.Set("newsize", defaultSizeString)
+
+	// Attempting to resize to the same size should return an error.
+	err = st.stdPostAPI("/host/storage/folders/resize", resizeValues)
+	if err == nil || err.Error() != storagemanager.ErrNoResize.Error() {
+		t.Fatalf("expected error %v, got %v", storagemanager.ErrNoResize, err)
+	}
+
+	// Try resizing to a bunch of sizes (invalid ones first, valid ones second).
+	// This ordering simplifies logic within the for loop.
+	for _, test := range resizeTests {
+		// Attempt to resize the host's storage folder.
+		resizeValues.Set("newsize", test.sizeString)
+		err = st.stdPostAPI("/host/storage/folders/resize", resizeValues)
+		if (err == nil && test.err != nil) || (err != nil && err.Error() != test.err.Error()) {
+			t.Fatalf("expected error to be %v, got %v", test.err, err)
+		}
+
+		// Find out if the resize call worked as expected.
+		if err := st.getAPI("/host/storage", &sg); err != nil {
+			t.Fatal(err)
+		}
+		// If the test size is valid, check that the folder has been resized
+		// properly.
+		if test.err == nil {
+			// Check that the folder's total capacity has been updated.
+			if got := sg.Folders[0].Capacity; got != test.size {
+				t.Fatalf("expected folder to be resized to %v; got %v instead", test.size, got)
+			}
+			// Check that the folder's remaining capacity has been updated.
+			if got := sg.Folders[0].CapacityRemaining; got != test.size {
+				t.Fatalf("folder should be empty, but capacity remaining (%v) != total capacity (%v)", got, test.size)
+			}
+		} else {
+			// If the test size is invalid, the folder should not have been
+			// resized. The invalid test cases are all run before the valid ones,
+			// so the folder size should still be defaultSize.
+			if got := sg.Folders[0].Capacity; got != defaultSize {
+				t.Fatalf("folder was resized to an invalid size (%v) in a test case that should have failed: %v", got, test)
+			}
+		}
+	}
+}
+
+// TestResizeNonemptyStorageFolder tests that invalid and valid calls to resize
+// a storage folder with one sector filled are properly handled.
+// Ideally, we would also test a very full storage folder (including the case
+// where the host tries to resize to a size smaller than the amount of data
+// in the folder), but that would be a very expensive test.
```

Comment:
If you have a storage folder that is twice as large as the minimum size allowed, it wouldn't be that expensive. The hardest part is probably that you need a renter to do the uploading, as there's no way in the API to upload data to the host without using a renter.

sectors during testing are set to 4kb.